Create boot menu entries when installed

TobiPeterG opened the pull request:
fixes #4
Tested on Tumbleweed, needs testing on MicroOS

lnussel commented:
That would create duplicate entries if called again though, right?

lnussel commented:
it does. also needs -p and -d options to tell it which disk, blindly defaults to sda1 which is not necessarily correct

TobiPeterG commented:
> it does. also needs -p and -d options to tell it which disk, blindly defaults to sda1 which is not necessarily correct

I added checks to only create the entries when there is none with the name already existing and added disk and partition number as arguments.
The function fails when the ESP couldn't be detected.

lnussel commented on sdbootutil:
```diff
@@ -669,13 +669,30 @@ install_sdboot()
 
 	mkdir -p "$boot_root/loader/entries"
 
+	esp_device=$(lsblk -o NAME,MOUNTPOINT -nl | grep '/boot/efi' | awk '{print $1}')
```
you don't need grep and awk, awk alone would do :-)
Anyway there's already `$arg_esp_path` parsed from bootctl. bootctl also lists the partition, something like `/dev/disk/by-partuuid/f0a6cf18-e241-b4ee-68c3-ecd933f49e82`. So should be possible to read that link. Then get the partition from sysfs like this: https://github.com/lnussel/addimageencryption/blob/master/addimageencryption#L234

TobiPeterG replied:
Ah I see, bootctl status also lists the ESP. arg_esp_path links to /boot/efi right? (so the path ralative to root)
I now use bootctl's listing, is this solution now better? :)

lnussel replied:
how about using `$arg_esp_path` directly?

Note just cutting off numbers to get rid of the partition is not sufficient. Think of `/dev/loop0p1`. You need to go through sysfs as pointed out in the link

TobiPeterG replied:
It should work now, I'm also handling the case where it points to an image, is this useful?

lnussel commented on sdbootutil:
```diff
@@ -669,13 +669,30 @@ install_sdboot()
+        efibootmgr | grep -q 'Boot.*openSUSE Secure Boot' || efibootmgr -q --create --disk $drive --part $partition_number --label "openSUSE Secure Boot" --loader "$sdboot_dst/shim.efi"
+
+        efibootmgr | grep -q 'Boot.*openSUSE Boot Manager' || efibootmgr -q --create --disk $drive --part $partition_number --label "openSUSE Boot Manager" --loader "$sdboot_dst/grub.efi"
+
```
should be sufficient to point at shim

lnussel replied:
also the path needs be the fat path, right? Ie `\EFI\systemd\shim.efi`
and don't forget to quote the variables. The shellcheck program is helpful for that.

TobiPeterG replied:
the path should be right, as least when I tested it, it worked :sweat_smile: 
I removed the grub entry and quoted the variables

lnussel replied:
systemd explicity turns around the slashes so there must be a reason for it. maybe some firmwares accept unix paths but that does not mean it's correct

TobiPeterG replied:
I have done some testing, efibootmgr automatically converts forwards slashes to backwards slashes.
This is also stated in the arch wiki:
https://wiki.archlinux.org/title/Unified_Extensible_Firmware_Interface#efibootmgr
`Note: UEFI uses backward slash \ as path separator but efibootmgr automatically converts UNIX-style / path separators.`

lnussel commented on sdbootutil:
```diff
@@ -718,13 +718,34 @@ install_sdboot()
 
+	if [ -d "$arg_esp_path" ] || [ -b "$arg_esp_path" ]; then
+	    if [ -b "$arg_esp_path" ]; then
+	        blkpart="$arg_esp_path"
```
is there reason to believe that arg_esp_path may refer to a block device?

lnussel commented on sdbootutil:
```diff
@@ -718,13 +718,34 @@ install_sdboot()
+	        mountpoint -q "$arg_esp_path" || { echo "$arg_esp_path is not a valid mountpoint"; return 1; }
+	        blkpart="$(findmnt -nvo SOURCE "$arg_esp_path")"
+	    fi
+	    [ -L "/sys/class/block/${blkpart##*/}" ] || { echo "$blkpart is not a partition"; return 1; }
```
I think you can just use err() here to make the tool exit

lnussel replied:
(and in the other error cases too)

lnussel commented:
merged, I'll adjust to my liking :)
